README
NAME
    Swagger2 - Swagger RESTful API Documentation

VERSION
    0.89

DEPRECATION WARNING
    The Swagger2 distribution is no longer actively maintained. Only severe
    bug fixes and pull requests will move this code forward. The reason
    behind this is that the code is too complex and hard to maintain.

    So what should you use instead?

    * Swagger2

      Swagger2 is either not very useful or replaced by JSON::Validator.

    * Swagger2::Client

      No alternatives. The issue with this module is that it does not
      understand if you have parameters with the same name. There might be a
      OpenAPI::Client at some point, but it is currently no plans to write
      it.

    * Swagger2::Editor

      No alternatives.

    * Swagger2::POD

      Swagger2::POD is not very good and also very hard to maintain.
      Mojolicious::Plugin::OpenAPI has a HTML renderer which makes
      documentation that is much easier to read and always in sync with the
      application.

      When that is said: The renderer in Mojolicious::Plugin::OpenAPI need
      refinement.

    * Swagger2::SchemaValidator

      Mojolicious::Plugin::OpenAPI has the validator built in. For other
      purposes, use JSON::Validator or JSON::Validator::OpenAPI instead.

    * Mojolicious::Command::swagger2

      No alternatives.

    * Mojolicious::Plugin::Swagger2

      Use Mojolicious::Plugin::OpenAPI instead. Mojolicious::Plugin::OpenAPI
      plays much nicer together with the Mojolicious framework.

DESCRIPTION
    Swagger2 is a module for generating, parsing and transforming swagger
    <http://swagger.io/> API specification. It has support for reading
    swagger specification in JSON notation and as well YAML format.

    Please read
    <http://thorsen.pm/perl/programming/2015/07/05/mojolicious-swagger2.html
    > for an introduction to Swagger and reasons for why you would to use
    it.

  Mojolicious server side code generator
    This distribution comes with a Mojolicious plugin,
    Mojolicious::Plugin::Swagger2, which can set up routes and perform input
    and output validation.

  Mojolicious client side code generator
    Swagger2 also comes with a Swagger2::Client generator, which converts
    the client spec to perl code in memory.

RECOMMENDED MODULES
    *   YAML parser

        A YAML parser is required if you want to read/write spec written in
        the YAML format. Supported modules are YAML::XS, YAML::Syck, YAML
        and YAML::Tiny.

SYNOPSIS
      use Swagger2;
      my $swagger = Swagger2->new("/path/to/api-spec.yaml");

      # Access the raw specification values
      print $swagger->api_spec->get("/swagger");

      # Returns the specification as a POD document
      print $swagger->pod->to_string;

ATTRIBUTES
  api_spec
      $pointer = $self->api_spec;
      $self = $self->api_spec(Mojo::JSON::Pointer->new({}));

    Holds a Mojo::JSON::Pointer object containing your API specification.

  base_url
      $mojo_url = $self->base_url;

    Mojo::URL object that holds the location to the API endpoint. Note: This
    might also just be a dummy URL to <http://example.com/>.

  ua
      $ua = $self->ua;
      $self = $self->ua(Mojo::UserAgent->new);

    A Mojo::UserAgent used to fetch remote documentation.

  url
      $mojo_url = $self->url;

    Mojo::URL object that holds the location to the documentation file. This
    can be both a location on disk or an URL to a server. A remote resource
    will be fetched using Mojo::UserAgent.

METHODS
  expand
      $swagger = $self->expand;

    This method returns a new "Swagger2" object, where all the references
    <https://tools.ietf.org/html/draft-zyp-json-schema-03#section-5.28> are
    resolved.

  find_operations
      $operations = $self->find_operations(\%q);

    Used to find a list of Operation Objects
    <http://swagger.io/specification/#operationObject> from the
    specification. %q can be:

      $all        = $self->find_operations;
      $operations = $self->find_operations($operationId);
      $operations = $self->find_operations({operationId => "listPets"});
      $operations = $self->find_operations({method => "post", path => "/pets"});
      $operations = $self->find_operations({tag => "pets"});

  javascript_client
      $file = $self->javascript_client;

    Returns a Mojo::Asset::File object which points to a file containing a
    custom JavaScript file which can communicate with
    Mojolicious::Plugin::Swagger2.

    See
    <https://github.com/jhthorsen/swagger2/blob/master/lib/Swagger2/swagger2
    -client.js> for source code.

    "swagger2-client.js" is currently EXPERIMENTAL!

  load
      $self = $self->load;
      $self = $self->load($url);

    Used to load the content from $url or "url". This method will try to
    guess the content type (JSON or YAML) by looking at the content of the
    $url.

  new
      $self = Swagger2->new($url);
      $self = Swagger2->new(%attributes);
      $self = Swagger2->new(\%attributes);

    Object constructor.

  parse
      $self = $self->parse($text);

    Used to parse $text instead of loading data from "url".

    The type of input text can be either JSON or YAML. It will default to
    YAML, but parse the text as JSON if it starts with "{".

  pod
      $pod_object = $self->pod;

    Returns a Swagger2::POD object.

  to_string
      $json = $self->to_string;
      $json = $self->to_string("json");
      $yaml = $self->to_string("yaml");

    This method can transform this object into Swagger spec.

  validate
      @errors = $self->validate;

    Will validate "api_spec" against Swagger RESTful API Documentation
    Specification
    <https://github.com/swagger-api/swagger-spec/blob/master/versions/2.0.md
    >, and return a list with all the errors found. See also "validate" in
    JSON::Validator.

COPYRIGHT AND LICENSE
    Copyright (C) 2014-2015, Jan Henning Thorsen

    This program is free software, you can redistribute it and/or modify it
    under the terms of the Artistic License version 2.0.

AUTHOR
    Jan Henning Thorsen - "jhthorsen@cpan.org"



Changes
Revision history for perl distribution Swagger2

0.89 2017-01-18T11:50:48+0000
 - Add skip_all to plugin-yaml.t to prevent random failures when used
 with JSON::Validator 0.91+

0.88 2016-12-24T14:35:48+0100
 - Bump required JSON::Validator version to 0.90
 - Add deprecation warning to Mojolicious::Plugin::Swagger2
 - Deleted guides, since this dist will be deprecated

0.87 2016-12-11T11:39:23+0100
 - Compatible with JSON::Validator 0.90

0.86 2016-09-19T18:55:09+0200
 - Deprecated distribution

0.85 2016-08-09T12:48:13+0200
 - Fix finding all $ref occurances in JSON::Validator #95

0.84 2016-08-08T10:04:28+0200
 - Fix recurring requests with same path part #93

0.83 2016-07-02T15:03:38+0200
 - Bump JSON::Validator dependency to 0.75
 - Change JSON::Validator will pass on uploads as Mojo::Upload objects

0.82 2016-06-22T19:27:24+0200
 - Deprecating Swagger2::SchemaValidator in favor of JSON::Validator::OpenAPI
 - Fix Swagger2::Client formData parameters #87
   Contributor: A.J. Lucas

0.81 2016-06-10T19:29:45-0700
 - Fix Swagger2::POD looping over "x-" HTTP methods
 - Change Swagger2::Client to set Content-Type on faked error responses #84
   Contributor: A.J. Lucas

0.80 2016-06-07T15:29:56-0500
 - Fix validating required "file" in formData
 - Add support for "discriminator" #72

0.79 2016-05-23T19:15:25+0200
 - Fix collectionFormat "multi"

0.78 2016-05-19T18:39:42+0200
 - Add EXPERIMENTAL .html and .pod handling for "spec_path" rendering

0.77 2016-04-12T08:32:56+0200
 - Add Swagger2->find_operations() method
 - Add support for plain WebSocket object passed to ws() in swagger2-client.js
 - Add http() to swagger2-client.js
 - Add load() to swagger2-client.js
 - Change ws() to be a setter and getter in swagger2-client.js
 - Mark "swagger_route_added" as DEPRECATED
 - Remove fresh() and cache features from swagger2-client.js

0.76 2016-04-09T09:52:34+0200
 - Fix new($file)->expand->validate #80
   Contributors: Tina Müller and Daniel Böhmer
 - Add EXPERIMENTAL support for SWAGGER2_IO_LOGGING=debug
 - Documented different response types #78 #79
 - Removed deprecated methods specification() and tree()

0.75 2016-02-22T09:02:43+0000
 - Fix indentation of JSONSchemaSupport guide

0.74 2016-02-19T11:28:52+0000
 - Add x-json-schema support
   Contributor: Martin Renvoize

0.73 2016-02-15T14:04:15+0100
 - Fix "0" as placeholder value #74
   Contributor: Le Garff Yoann

0.72 2016-02-12T11:30:05+0100
 - Fix rendering Swagger spec cannot contain "id" and "x-mojo-xxx"
 - Fix rendering Swagger spec with requested host in response.
 - Made the code less confusing by removing the /o regex modifier #71
   Contributor: Daniel Böhmer

0.71 2016-02-08T21:29:57+0100
 - Bundle EXPERIMENTAL swagger2-client.js
 - Fix spelling error #70
   Contributor: Lucas Kanashiro

0.70 2016-01-29T10:50:57+0100
 - Documentation fixes
   Contributor: Bernhard Graf
 - Remove outdated warning #69
   Contributor: Denny de la Haye

0.69 2016-01-14T22:01:15+0100
 - Allow blocking Swagger2::Client requests to return on failure #67

0.68 2016-01-13T16:36:28+0100
 - Can serve specification from basePath
 - Fix skipping tests on MSWin32 #66
   Contributor: Paul Durden

0.67 2016-01-12T14:36:24+0100
 - Add "basePath" to "Add route ..." debug message #64
 - Can pass Swagger2->new to Swagger2::Client #63
 - Fix new(url => Mojo::URL->new(...));
 - Fix expand() will not clear url()

0.66 2015-12-29T11:52:06+0100
 - Will not add non-existing input keys to $args #52
   Contributor: Scott Wiersdorf
 - Use $c->match instead of $c->stash to get placeholder data
 - Add "From" to list of ignored words in SKIP_OP_RE #54
 - Allow more complex method name detection from operationId #55
 - Skipping editor test on Win32 #56
 - Boolean in headers will be the strings "true" and "false" #58

0.65 2015-12-03T12:51:51+0100
 - Swagger2::Client will generate both camel- and snake case methods
 - Swagger2::Client pass $err as a string to callback
 - Swagger2::Editor can save specification back to disk on POST

0.64 2015-11-28T08:18:45+0100
 - YAML::Tiny is no longer supported by JSON::Validator
 - Bump JSON::Validator dependency to 0.63

0.63 2015-11-19T09:22:44+0100
 - Add validation of output headers, fixes #41
 - Add support for "readOnly" in object properties #49

0.62 2015-11-04T12:42:32+0000
 - Trying to authorise release

0.61 2015-11-03T20:22:55+0000
 - Fix handling of empty collectionFormat value
 - Fix handling of arrays in request body
 - Add validation for 204 No Content response

0.60 2015-10-14T23:25:14+0200
 - Fix pass on $app to WebSocket controller
 - Fix support for "collectionFormat" for both "parameters" and "items"
 - Add support for "collectionFormat" "multi"

0.59 2015-10-11T13:22:31+0200
 - Fix validating required body
 - Add resource to download full API spec
 - Add before_render hook to make sure we reply with JSON
 - Add EXPERIMENTAL dispatch_to_swagger()

0.58 2015-10-06T11:36:29+0200
 - Add support for using operationId instead of x-mojo-controller #36 #37
 - Add "swagger_route_added" event #27
 - Remove "valid" from standard error structure since it was redundant
 - Documented how to include http://git.io/vcKD4

0.57 2015-09-30T11:48:22+0200
 - Require JSON::Validator 0.56 for YAML validation
 - Add support for passing "coerce" to M::P::Swagger2::register()

0.56 2015-09-29T20:23:07+0200
 - Add Mojolicious::Plugin::Swagger2 will validate spec on load

0.55 2015-09-27T16:53:08+0200
 - Split Mojolicious::Plugin::Swagger2 into guides
 - Remove EXPERIMENTAL from "x-mojo-around-action"

0.54 2015-09-27T13:37:06+0200
 - Compatible with JSON::Validator 0.54
 - Compatible with Mojolicious 6.22
 - Rename index.html.ep to list.html.ep in blog example #28
   Contributor: Stephan Jauernick
 - Paths are sorted from the most specific to least specific #35
   Contributor: Olli-Antti Kivilahti

0.53 2015-09-13T15:55:54+0200
 - Add EXPERIMENTAL naming of swagger routes #27
 - Add "swagger_operation_spec" to route defaults #30

Makefile.PL
# Generated by git-ship. See 'git-ship --man' for help or https://github.com/jhthorsen/app-git-ship
use ExtUtils::MakeMaker;
WriteMakefile(
  NAME          => 'Swagger2',
  AUTHOR        => 'Martin Renvoize <martin.renvoize@ptfs-europe.com>',
  LICENSE       => 'artistic_2',
  ABSTRACT_FROM => 'lib/Swagger2.pm',
  VERSION_FROM  => 'lib/Swagger2.pm',
  EXE_FILES     => [qw(  )],
  META_MERGE    => {
    resources => {
      bugtracker => 'https://github.com/jhthorsen/swagger2/issues',
      homepage   => 'https://github.com/jhthorsen/swagger2',
      repository => 'https://github.com/mrenvoize/swagger2.git',
    },
  },
  BUILD_REQUIRES => {'Test::More'      => '0.88'},
  PREREQ_PM      => {'JSON::Validator' => '0.90', 'Mojolicious' => '6.00'},
  test           => {TESTS             => 't/*.t'},
);


lib/Swagger2.pm
package Swagger2;
use Mojo::Base -base;
use Mojo::Asset::File;
use Mojo::JSON;
use Mojo::JSON::Pointer;
use Mojo::URL;
use File::Basename ();
use File::Spec;
use JSON::Validator::OpenAPI;

our $VERSION = '0.89';

# Should be considered internal
our $JS_CLIENT
  = File::Spec->catfile(File::Basename::dirname(__FILE__), 'Swagger2', 'swagger2-client.js');

has api_spec => sub {
  my $self = shift;
  return $self->_validator->_load_schema($self->url) if '' . $self->url;
  return Mojo::JSON::Pointer->new({});
};

has base_url => sub {
  my $self = shift;
  my $url  = Mojo::URL->new;
  my ($schemes, $v);

  $self->load if !$self->{api_spec} and '' . $self->url;
  $schemes = $self->api_spec->get('/schemes') || [];
  $url->host($self->api_spec->get('/host')     || 'example.com');
  $url->path($self->api_spec->get('/basePath') || '/');
  $url->scheme($schemes->[0]                   || 'http');

  return $url;
};

has _specification => sub { shift->_validator->schema('http://swagger.io/v2/schema.json')->schema };

has _validator => sub { JSON::Validator::OpenAPI->new };

sub ua  { shift->_validator->ua(@_) }
sub url { shift->{url} }

sub expand {
  my $self   = shift;
  my $class  = Scalar::Util::blessed($self);
  my $schema = $self->_validator->schema($self->api_spec->data)->schema;
  $class->new(%$self)->api_spec($schema);
}

sub find_operations {
  my ($self, $needle) = @_;
  my $paths      = $self->api_spec->get('/paths');
  my $operations = [];

  $needle ||= {};
  $needle = {operationId => $needle} unless ref $needle;

  for my $path (keys %$paths) {
    next if $path =~ /^x-/;
    next if $needle->{path} and $needle->{path} ne $path;
    for my $method (keys %{$paths->{$path}}) {
      my $object = $paths->{$path}{$method};
      next if $method =~ /^x-/;
      next if $needle->{tag} and !grep { $needle->{tag} eq $_ } @{$object->{tags} || []};
      next if $needle->{method} and $needle->{method} ne $method;
      next if $needle->{operationId} and $needle->{operationId} ne $object->{operationId};
      push @$operations, $object;
    }
  }

  return $operations;
}

sub javascript_client { Mojo::Asset::File->new(path => $JS_CLIENT) }

sub load {
  my $self = shift;
  delete $self->{base_url};
  $self->{url} = Mojo::URL->new(shift) if @_;
  $self->{api_spec} = $self->api_spec;
  $self;
}

sub new {
  my $class = shift;
  my $url   = @_ % 2 ? shift : '';
  my $self  = $class->SUPER::new(@_);

  $url =~ s!^file://!!;
  $self->{url} ||= $url;
  $self->{url} = Mojo::URL->new($self->{url}) unless ref $self->{url};
  $self;
}

sub parse {
  my ($self, $doc, $namespace) = @_;
  delete $self->{base_url};
  $namespace ||= 'http://127.0.0.1/#';
  $self->{url}      = Mojo::URL->new($namespace);
  $self->{api_spec} = Mojo::JSON::Pointer->new($self->_validator->_load_schema_from_text($doc));
  $self;
}

sub pod {
  my $self     = shift;
  my $resolved = $self->_validator->schema($self->api_spec->data)->schema;
  require Swagger2::POD;
  Swagger2::POD->new(base_url => $self->base_url, api_spec => $resolved);
}

sub to_string {
  my $self = shift;
  my $format = shift || 'json';

  return DumpYAML($self->api_spec->data) if $format eq 'yaml';
  return Mojo::JSON::encode_json($self->api_spec->data);
}

sub validate {
  my $self = shift;
  $self->_validator->validate($self->expand->api_spec->data, $self->_specification->data);
}

sub _is_true {
  return $_[0] if ref $_[0] and !Scalar::Util::blessed($_[0]);
  return 0 if !$_[0] or $_[0] =~ /^(n|false|off)/i;
  return 1;
}

1;

=encoding utf8

=head1 NAME

Swagger2 - Swagger RESTful API Documentation

=head1 VERSION

0.89

=head1 DEPRECATION WARNING

The L<Swagger2> distribution is no longer actively maintained. Only severe bug
fixes and pull requests will move this code forward.  The reason behind this is
that the code is too complex and hard to maintain.

So what should you use instead?

=over 2

=item * L<Swagger2>

L<Swagger2> is either not very useful or replaced by L<JSON::Validator>.

=item * L<Swagger2::Client>

No alternatives. The issue with this module is that it does not understand if
you have parameters with the same name. There might be a L<OpenAPI::Client> at
some point, but it is currently no plans to write it.

=item * L<Swagger2::Editor>

No alternatives.

=item * L<Swagger2::POD>

L<Swagger2::POD> is not very good and also very hard to maintain.
L<Mojolicious::Plugin::OpenAPI> has a HTML renderer which makes documentation
that is much easier to read and always in sync with the application.

When that is said: The renderer in L<Mojolicious::Plugin::OpenAPI> need
refinement.

=item * L<Swagger2::SchemaValidator>

L<Mojolicious::Plugin::OpenAPI> has the validator built in. For other purposes,
use L<JSON::Validator> or L<JSON::Validator::OpenAPI> instead.

=item * L<Mojolicious::Command::swagger2>

No alternatives.

=item * L<Mojolicious::Plugin::Swagger2>

Use L<Mojolicious::Plugin::OpenAPI> instead. L<Mojolicious::Plugin::OpenAPI>
plays much nicer together with the L<Mojolicious> framework.

=back

=head1 DESCRIPTION

L<Swagger2> is a module for generating, parsing and transforming
L<swagger|http://swagger.io/> API specification. It has support for reading
swagger specification in JSON notation and as well YAML format.

Please read L<http://thorsen.pm/perl/programming/2015/07/05/mojolicious-swagger2.html>
for an introduction to Swagger and reasons for why you would to use it.

=head2 Mojolicious server side code generator

This distribution comes with a L<Mojolicious> plugin,
L<Mojolicious::Plugin::Swagger2>, which can set up routes and perform input
and output validation.

=head2 Mojolicious client side code generator

Swagger2 also comes with a L<Swagger2::Client> generator, which converts the client
spec to perl code in memory.

=head1 RECOMMENDED MODULES

=over 4

=item * YAML parser

A L<YAML> parser is required if you want to read/write spec written in
the YAML format. Supported modules are L<YAML::XS>, L<YAML::Syck>, L<YAML>
and L<YAML::Tiny>.

=back

=head1 SYNOPSIS

  use Swagger2;
  my $swagger = Swagger2->new("/path/to/api-spec.yaml");

  # Access the raw specification values
  print $swagger->api_spec->get("/swagger");

  # Returns the specification as a POD document
  print $swagger->pod->to_string;

=head1 ATTRIBUTES

=head2 api_spec

  $pointer = $self->api_spec;
  $self = $self->api_spec(Mojo::JSON::Pointer->new({}));

Holds a L<Mojo::JSON::Pointer> object containing your API specification.

=head2 base_url

  $mojo_url = $self->base_url;

L<Mojo::URL> object that holds the location to the API endpoint.
Note: This might also just be a dummy URL to L<http://example.com/>.

=head2 ua

  $ua = $self->ua;
  $self = $self->ua(Mojo::UserAgent->new);

A L<Mojo::UserAgent> used to fetch remote documentation.

=head2 url

  $mojo_url = $self->url;

L<Mojo::URL> object that holds the location to the documentation file.
This can be both a location on disk or an URL to a server. A remote
resource will be fetched using L<Mojo::UserAgent>.

=head1 METHODS

=head2 expand

  $swagger = $self->expand;

This method returns a new C<Swagger2> object, where all the
L<references|https://tools.ietf.org/html/draft-zyp-json-schema-03#section-5.28>
are resolved.

=head2 find_operations

  $operations = $self->find_operations(\%q);

Used to find a list of L<Operation Objects|http://swagger.io/specification/#operationObject>
from the specification. C<%q> can be:

  $all        = $self->find_operations;
  $operations = $self->find_operations($operationId);
  $operations = $self->find_operations({operationId => "listPets"});
  $operations = $self->find_operations({method => "post", path => "/pets"});
  $operations = $self->find_operations({tag => "pets"});

=head2 javascript_client

  $file = $self->javascript_client;

Returns a L<Mojo::Asset::File> object which points to a file containing a
custom JavaScript file which can communicate with
L<Mojolicious::Plugin::Swagger2>.

See L<https://github.com/jhthorsen/swagger2/blob/master/lib/Swagger2/swagger2-client.js>
for source code.

C<swagger2-client.js> is currently EXPERIMENTAL!

=head2 load

  $self = $self->load;
  $self = $self->load($url);

Used to load the content from C<$url> or L</url>. This method will try to
guess the content type (JSON or YAML) by looking at the content of the C<$url>.

=head2 new

  $self = Swagger2->new($url);
  $self = Swagger2->new(%attributes);
  $self = Swagger2->new(\%attributes);

Object constructor.

=head2 parse

  $self = $self->parse($text);

Used to parse C<$text> instead of L<loading|/load> data from L</url>.

The type of input text can be either JSON or YAML. It will default to YAML,
but parse the text as JSON if it starts with "{".

=head2 pod

  $pod_object = $self->pod;

Returns a L<Swagger2::POD> object.

=head2 to_string

  $json = $self->to_string;
  $json = $self->to_string("json");
  $yaml = $self->to_string("yaml");

This method can transform this object into Swagger spec.

=head2 validate

  @errors = $self->validate;

Will validate L</api_spec> against
L<Swagger RESTful API Documentation Specification|https://github.com/swagger-api/swagger-spec/blob/master/versions/2.0.md>,
and return a list with all the errors found. See also L<JSON::Validator/validate>.

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014-2015, Jan Henning Thorsen

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

=head1 AUTHOR

Jan Henning Thorsen - C<jhthorsen@cpan.org>

=cut
